File: app/src/main/java/com/ue/colorful/feature/game/TimeTrialModeActivity.java
package com.ue.colorful.feature.game;

import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.ue.colorful.R;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class TimeTrialModeActivity extends AppCompatActivity implements View.OnClickListener {
    private static final int MEDIUM = 2;
    private static final int HARD = 3;

    private TextView tv_level;
    private Chronometer chronometer;
    private Button btn_redraw, btn_half;

    private int buttons_in_row;
    private int random_button;
    private int width;
    private int level;
    private LinearLayout layout;

    private Arrays arrays = new Arrays();
    private Random r = new Random();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_time_trial_mode);

        chronometer = (Chronometer) findViewById(R.id.chronometer1);
        tv_level = (TextView) findViewById(R.id.tv_level);
        tv_level.setText("10");
        btn_redraw = (Button) findViewById(R.id.btn_redraw);
        btn_half = (Button) findViewById(R.id.btn_half);

        layout = (LinearLayout) findViewById(R.id.linear_layout_tags);

        btn_redraw.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                layout.removeAllViews();
                drawMap(HARD, 7);

            }
        });

        btn_half.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                halfTiles(49);
            }
        });
        startGame();
    }

    private void drawMap(int level, int buttons) {
        String color0, color1;

        buttons_in_row = buttons;
        random_button = (r.nextInt(buttons_in_row * buttons_in_row) + 1);

        if (level == MEDIUM) {
            int random_color = r.nextInt(arrays.colors_medium.length);
            color0 = arrays.getMediumColor0(random_color);
            color1 = arrays.getMediumColor1(random_color);
        } else {
            int random_color = r.nextInt(arrays.colors_hard.length);
            color0 = arrays.getHardColor0(random_color);
            color1 = arrays.getHardColor1(random_color);
        }

        for (int i = 0; i < buttons_in_row; i++) {
            LinearLayout row = new LinearLayout(this);
            row.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));

            for (int j = 0; j < buttons_in_row; j++) {
                Button btn = new Button(this);
                LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT, 1.0f);
                params.setMargins(5, 5, 5, 5);
                btn.setLayoutParams(params);
                btn.setId(j + 1 + (i * buttons_in_row));
                btn.setWidth(width / buttons_in_row);
                btn.setHeight(width / buttons_in_row);
                btn.setBackgroundResource(R.drawable.button_wrong);
                GradientDrawable drawable = (GradientDrawable) btn.getBackground();
                drawable.setColor(Color.parseColor("#" + color0));
                btn.setOnClickListener(this);
                row.addView(btn);
            }

            layout.addView(row);
        }
        Button b = (Button) layout.findViewById(random_button);
        GradientDrawable drawable2 = (GradientDrawable) b.getBackground();
        drawable2.setColor(Color.parseColor("#" + color1));
    }

    public void onClick(View view) {
        ((Button) view).setText("*");
        view.setEnabled(false);

        int myId = view.getId();
        if (myId == random_button) {

            layout.removeAllViews();

            level--;
            if (level > 7) {
                drawMap(MEDIUM, 7);
            } else if (level > 0 && level <= 7) {
                drawMap(HARD, 7);
            } else {
                gameOver();
            }

            tv_level.setText(String.valueOf(level));
        } else {
            Button b3 = (Button) layout.findViewById(myId);
            GradientDrawable drawable3 = (GradientDrawable) b3.getBackground();
            drawable3.setColor(Color.BLACK);
        }

    }

    public void halfTiles(int num) {

        Random r = new Random();
        int random_tile;

        List<Integer> list = new ArrayList<>();
        for (int i = 1; i <= num; i++) {
            list.add(i);
        }

        int x = list.size() / 2 + 1;

        List<Integer> list2 = new ArrayList<>();
        for (int i = 1; i < x; i++) {
            random_tile = (r.nextInt(list.size()) + 1);
            if (random_tile == random_button) {
                i--;
            } else if (!list2.contains(random_tile)) {
                list2.add(random_tile);
            } else {
                i--;
            }
        }

        for (int i = 0; i < list.size(); i++) {
            for (int j = 0; j < list2.size(); j++) {
                if (list.get(i) == list2.get(j)) {
                    Button b2 = (Button) layout.findViewById(list.get(i));
                    b2.setVisibility(View.INVISIBLE);
                }
            }
        }
    }

    public void gameOver() {
        if (chronometer != null) chronometer.stop();

        TimeTrialResultDialog dialog = TimeTrialResultDialog.newInstance(chronometer.getText().toString());
        dialog.setPlayAgainListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startGame();
            }
        });
        dialog.show(getSupportFragmentManager(), "");
    }

    public void startGame() {
        if (chronometer != null) chronometer.start();

        width = (getResources().getDisplayMetrics().widthPixels * 9) / 10;
        layout.removeAllViews();

        level = 10;
        tv_level.setText(String.valueOf(level));
        drawMap(HARD, 7);
    }

    @Override
    public void onPause() {
        super.onPause();
        Log.i("onononT", "onPause");
        //if user opens another app then we need to set this otherwise Connected to Google+ msg will be displayed
        if (chronometer != null) chronometer.stop();
    }

    @Override
    public void onStop() {
        super.onStop();
        Log.i("onononT", "onStop");
        if (chronometer != null) chronometer.stop();
    }

    @Override
    public void onResume() {
        super.onResume();
        Log.i("onononT", "onResume");
        //for now when the user comes back from another app we don't stop the clock
        if (chronometer.getTimeElapsed() > 0) {
            chronometer.resume();
            //Log.i("getTimeElapsed", String.valueOf(chronometer.getTimeElapsed())); //11178 = 11 s 178 ms, 8301 = 8s301ms
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (chronometer != null) chronometer.stop();
    }
}

File: app/src/main/java/com/ue/colorful/feature/game/TimeTrialModeActivity.kt
package com.ue.colorful.feature.game

import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.Button
import android.widget.LinearLayout
import com.ue.colorful.R
import kotlinx.android.synthetic.main.activity_time_trial_mode.*
import java.util.*

class TimeTrialModeActivity : AppCompatActivity(), View.OnClickListener {
    private var buttonsInRow: Int = 0
    private var randomButton: Int = 0
    private var width: Int = 0
    private var level: Int = 0

    private val arrays = Arrays()
    private val r = Random()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_time_trial_mode)

        tvLevel.text = "10"

        btnRedraw.setOnClickListener {
            linearLayoutTags.removeAllViews()
            drawMap(HARD, 7)
        }

        btnHalf.setOnClickListener { halfTiles(49) }
        startGame()
    }

    private fun drawMap(level: Int, buttons: Int) {
        val color0: String
        val color1: String

        buttonsInRow = buttons
        randomButton = r.nextInt(buttonsInRow * buttonsInRow) + 1

        when (level) {
            MEDIUM -> {
                val randomColor = r.nextInt(arrays.mediumColors.size)
                color0 = arrays.getMediumColor0(randomColor)
                color1 = arrays.getMediumColor1(randomColor)
            }
            else -> {
                val randomColor = r.nextInt(arrays.hardColors.size)
                color0 = arrays.getHardColor0(randomColor)
                color1 = arrays.getHardColor1(randomColor)
            }
        }

        for (i in 0 until buttonsInRow) {
            val row = LinearLayout(this)
            row.layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)

            for (j in 0 until buttonsInRow) {
                val btn = Button(this)
                val params = LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT, 1.0f)
                params.setMargins(5, 5, 5, 5)
                btn.layoutParams = params
                btn.id = j + 1 + i * buttonsInRow
                btn.width = width / buttonsInRow
                btn.height = width / buttonsInRow
                btn.setBackgroundResource(R.drawable.button_wrong)
                val drawable = btn.background as GradientDrawable
                drawable.setColor(Color.parseColor("#" + color0))
                btn.setOnClickListener(this)
                row.addView(btn)
            }

            linearLayoutTags.addView(row)
        }
        val b = linearLayoutTags.findViewById<View>(randomButton) as Button
        val drawable2 = b.background as GradientDrawable
        drawable2.setColor(Color.parseColor("#" + color1))
    }

    override fun onClick(view: View) {
        (view as Button).text = "*"
        view.setEnabled(false)

        val myId = view.getId()
        if (myId == randomButton) {
            linearLayoutTags.removeAllViews()

            level--
            if (level > 7) {
                drawMap(MEDIUM, 7)
            } else if (level > 0 && level <= 7) {
                drawMap(HARD, 7)
            } else {
                gameOver()
            }

            tvLevel.text = level.toString()
        } else {
            val b3 = linearLayoutTags.findViewById<View>(myId) as Button
            val drawable3 = b3.background as GradientDrawable
            drawable3.setColor(Color.BLACK)
        }
    }

    fun halfTiles(num: Int) {
        val r = Random()
        var random_tile: Int

        val list = ArrayList<Int>()
        for (i in 1..num) {
            list.add(i)
        }

        val x = list.size / 2 + 1

        val list2 = ArrayList<Int>()
        run {
            var i = 1
            while (i < x) {
                random_tile = r.nextInt(list.size) + 1
                if (random_tile == randomButton) {
                    i--
                } else if (!list2.contains(random_tile)) {
                    list2.add(random_tile)
                } else {
                    i--
                }
                i++
            }
        }

        for (i in list.indices) {
            for (j in list2.indices) {
                if (list[i] === list2[j]) {
                    val b2 = linearLayoutTags.findViewById<View>(list[i]) as Button
                    b2.visibility = View.INVISIBLE
                }
            }
        }
    }

    fun gameOver() {
        chronometerText.stop()

        val dialog = TimeTrialResultDialog.newInstance(chronometerText.text.toString())
        dialog.setPlayAgainListener(View.OnClickListener { startGame() })
        dialog.show(supportFragmentManager, "")
    }

    fun startGame() {
        chronometerText.start()

        width = resources.displayMetrics.widthPixels * 9 / 10
        linearLayoutTags.removeAllViews()

        level = 10
        tvLevel.text = level.toString()
        drawMap(HARD, 7)
    }

    public override fun onPause() {
        super.onPause()
        chronometerText.stop()
    }

    public override fun onStop() {
        super.onStop()
        chronometerText.stop()
    }

    public override fun onResume() {
        super.onResume()
        if (chronometerText.timeElapsed > 0) {
            chronometerText.resume()
        }
    }

    public override fun onDestroy() {
        super.onDestroy()
        chronometerText.stop()
    }

    companion object {
        private val MEDIUM = 2
        private val HARD = 3
    }
}
